massa-ledger-exports/src/key.rs

```diff
@@ -1,105 +1,191 @@
 use massa_models::{
-    address::{Address, AddressDeserializer, ADDRESS_SIZE_BYTES},
+    address::{Address, AddressDeserializer, AddressSerializer},
     serialization::{VecU8Deserializer, VecU8Serializer},
 };
-use massa_serialization::{DeserializeError, Deserializer, SerializeError, Serializer};
+use massa_serialization::{
+    Deserializer, SerializeError, Serializer, U64VarIntDeserializer, U64VarIntSerializer,
+};
 use nom::error::{ContextError, ParseError};
+use num_enum::{IntoPrimitive, TryFromPrimitive};
 use std::ops::Bound::Included;
 
 pub const BALANCE_IDENT: u8 = 0u8;
 pub const BYTECODE_IDENT: u8 = 1u8;
 pub const DATASTORE_IDENT: u8 = 2u8;
+pub const KEY_VERSION: u64 = 0;
+
+#[derive(PartialEq, Eq, Clone, IntoPrimitive, TryFromPrimitive, Debug)]
+#[repr(u8)]
+enum KeyTypeId {
+    Balance = 0,
+    Bytecode = 1,
+    Datastore = 2,
+}
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum KeyType {
+    BALANCE,
+    BYTECODE,
+    DATASTORE(Vec<u8>),
+}
+
+#[derive(Default, Clone)]
+pub struct KeyTypeSerializer {
+    vec_u8_serializer: VecU8Serializer,
+    // Whether is deserialized with VecU8Deserializer or not.
+    // If true, we use the VecU8Serializer to serialize the key which will add the length at the beginning.
+    // If false, we just serialize the key as is.
+    // This allows us to store the datastore key length at the beginning of the key or not.
+    // The datastore key length is useful when transfering multiple keys, like in packets,
+    // but isn't when storing a datastore key in the ledger.
+    with_datastore_key_length: bool,
+}
+
+impl KeyTypeSerializer {
+    /// Creates a new KeyTypeSerializer.
+    /// `with_datastore_key_length` if true, the datastore key is serialized with its length.
+    pub fn new(with_datastore_key_length: bool) -> Self {
+        Self {
+            vec_u8_serializer: VecU8Serializer::new(),
+            with_datastore_key_length,
+        }
+    }
+}
+
+impl Serializer<KeyType> for KeyTypeSerializer {
+    fn serialize(&self, value: &KeyType, buffer: &mut Vec<u8>) -> Result<(), SerializeError> {
+        match value {
+            KeyType::BALANCE => buffer.extend(&[u8::from(KeyTypeId::Balance)]),
+            KeyType::BYTECODE => buffer.extend(&[u8::from(KeyTypeId::Bytecode)]),
+            KeyType::DATASTORE(data) => {
+                buffer.extend(&[u8::from(KeyTypeId::Datastore)]);
+                if self.with_datastore_key_length {
```

Comment:
I would vote to make this uniform everywhere

Reply:
this needs a followup issue

Reply:
#3708